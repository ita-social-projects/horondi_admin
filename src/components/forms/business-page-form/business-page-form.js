import React, { useEffect, useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Paper, TextField, Grid, Typography } from '@material-ui/core';
import { useFormik } from 'formik';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router';

import * as Yup from 'yup';
import useBusinessHandlers from '../../../utils/use-business-handlers';
import { useStyles } from './business-page-form.styles';
import { SaveButton, BackButton } from '../../buttons';
import LoadingBar from '../../loading-bar';

import {
  setCodeHandler,
  uaSetTitleHandler,
  uaSetTextHandler,
  enSetTitleHandler,
  enSetTextHandler,
  businessPageDispatchHandler,
  indexFinder
} from '../../../utils/business-page-form';

import {
  addBusinessPage,
  getBusinessPageById,
  updateBusinessPage
} from '../../../redux/business-pages/business-pages.actions';

import { useCommonStyles } from '../../../pages/common.styles';
import LanguagePanel from '../language-panel';
import { config } from '../../../configs';
import { useUnsavedChangesHandler } from '../../../hooks/form-dialog/use-unsaved-changes-handler';

const BusinessPageForm = ({ id, editMode }) => {
  const dispatch = useDispatch();
  const { loading, businessPage } = useSelector(({ BusinessPages }) => ({
    loading: BusinessPages.loading,
    businessPage: BusinessPages.currentPage
  }));

  const classes = useStyles();
  const common = useCommonStyles();
  const {
    labels: { businessPageLabel },
    languages,
    businessPageErrorMessages: {
      ENTER_CODE_ERROR_MESSAGE,
      ENTER_TITLE_ERROR_MESSAGE
    }
  } = config;

  const { pathToBusinessPages } = config.routes;

  const {
    createBusinessPage,
    uaSetText,
    enSetText,
    uaSetTitle,
    enSetTitle,
    uaText,
    enText,
    enTitle,
    uaTitle,
    code,
    setCode,
    files,
    setFiles
  } = useBusinessHandlers();

  useEffect(() => {
    id && dispatch(getBusinessPageById(id));
  }, [dispatch, id]);

  useEffect(() => {
    const isEditingReady = businessPage && editMode;

    setCode(setCodeHandler(isEditingReady, businessPage));
    uaSetTitle(uaSetTitleHandler(isEditingReady, businessPage));
    uaSetText(uaSetTextHandler(isEditingReady, businessPage));
    enSetTitle(enSetTitleHandler(isEditingReady, businessPage));
    enSetText(enSetTextHandler(isEditingReady, businessPage));
  }, [
    code,
    setCode,
    editMode,
    businessPage,
    uaSetText,
    uaSetTitle,
    enSetText,
    enSetTitle
  ]);

  const formSchema = Yup.object().shape({
    code: Yup.string().required(ENTER_CODE_ERROR_MESSAGE),
    uaTitle: Yup.string().required(ENTER_TITLE_ERROR_MESSAGE),
    enTitle: Yup.string().required(ENTER_TITLE_ERROR_MESSAGE)
  });

  const { values, errors, touched, handleSubmit, handleBlur, handleChange } =
    useFormik({
      initialValues: {
        code,
        uaTitle,
        enTitle,
        uaText,
        enText
      },
      validationSchema: formSchema,
      onSubmit: async () => {
        const uniqueFiles = files.filter((file, i) => {
          const { name, size } = file;
          return indexFinder(i, files, name, size);
        });

        const newUaText = values.uaText.replace(
          /src="data:image.*?"/g,
          'src=""'
        );
        const newEnText = values.enText.replace(
          /src="data:image.*?"/g,
          'src=""'
        );

        const page = createBusinessPage({
          ...values,
          uaText: newUaText,
          enText: newEnText
        });

        businessPageDispatchHandler(
          editMode,
          dispatch,
          updateBusinessPage,
          addBusinessPage,
          { id, page, files: uniqueFiles },
          { page, files: uniqueFiles }
        );
      }
    });

  useUnsavedChangesHandler(values);
  useMemo(() => {
    values.code = code;
    values.uaTitle = uaTitle;
    values.enTitle = enTitle;
    values.uaText = uaText;
    values.enText = enText;
  }, [code, uaTitle, enTitle, uaText, enText]);

  if (loading) {
    return <LoadingBar />;
  }

  businessPageLabel[1].setFiles = setFiles;

  const inputOptions = {
    errors,
    touched,
    handleChange,
    handleBlur,
    values,
    inputs: businessPageLabel
  };

  const eventPreventHandler = (e) => {
    e.preventDefault();
  };

  return (
    <div className={classes.container}>
      <div className={classes.buttonContainer}>
        <Grid container spacing={2} className={classes.fixedButtons}>
          <Grid item className={classes.button}>
            <BackButton pathBack={pathToBusinessPages} />
          </Grid>
          <Grid item className={classes.button}>
            <SaveButton
              id='save'
              type='submit'
              title='Зберегти'
              data-cy='save-btn'
              onClickHandler={handleSubmit}
              values={{
                code: values.code,
                uaTitle: values.uaTitle,
                enTitle: values.enTitle
              }}
              errors={errors}
            />
          </Grid>
        </Grid>
      </div>
      <div className={common.adminHeader}>
        <Typography
          variant='h1'
          className={common.materialTitle}
          data-cy='add-header'
        >
          {config.titles.businessPageTitles.addBusinessPageTitle}
        </Typography>
      </div>

      <form onSubmit={(e) => eventPreventHandler(e)}>
        <Grid item xs={12}>
          <Paper className={classes.businessPageForm}>
            <TextField
              id='code'
              className={classes.textField}
              variant='outlined'
              label='Код сторінки'
              value={values.code}
              onChange={handleChange}
              onBlur={handleBlur}
              error={touched.code && !!errors.code}
              data-cy='page-code'
            />
          </Paper>
          {touched.code && errors.code && (
            <div data-cy='code-error' className={classes.errorMessage}>
              {errors.code}
            </div>
          )}
        </Grid>
        {languages.map((lang) => (
          <LanguagePanel lang={lang} inputOptions={inputOptions} key={lang} />
        ))}
      </form>
    </div>
  );
};

BusinessPageForm.propTypes = {
  editMode: PropTypes.bool,
  id: PropTypes.string
};

BusinessPageForm.defaultProps = {
  editMode: false,
  id: null
};

export default withRouter(BusinessPageForm);
