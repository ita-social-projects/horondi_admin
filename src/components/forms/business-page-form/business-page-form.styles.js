import { makeStyles } from '@material-ui/core/styles';

export const useStyles = makeStyles((theme) => ({
  businessPageForm: {
    display: 'flex',
    flexDirection: 'column',
    margin: '20px 0',
    padding: '10px'
  },
  textField: {
    margin: '10px 5px'
  },
  tabs: {
    backgroundColor: '#424242'
  },
  controlsBlock: {
    display: 'flex',
    marginTop: 20
  },
  container: {
    padding: theme.spacing(2.5),
    width: '100%',
    position: 'relative',
    '@media (max-width: 481px)': {
      marginTop: '6px'
    }
  },
  buttonContainer: {
    position: 'relative',
    width: '100%',
    height: 60,
    marginTop: 51,
    '@media (max-width: 600px)': {
      marginTop: '36px'
    }
  },
  fixedButtons: {
    position: 'fixed',
    height: 60,
    zIndex: 1001,
    backgroundColor: theme.palette.bodyColor
  },
  button: {
    marginTop: theme.spacing(0.5)
  },
  errorMessage: {
    color: 'red',
    marginTop: 5,
    marginLeft: 15,
    fontSize: '.9em'
  },
  tabField: {
    padding: 20,
    display: 'flex',
    flexDirection: 'column'
  }
}));
