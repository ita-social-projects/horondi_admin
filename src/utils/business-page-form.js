export const setCodeHandler = (condition, businessPage) =>
  condition ? businessPage.code : '';

export const uaSetTitleHandler = (condition, businessPage) =>
  condition ? businessPage.title[0].value : '';

export const uaSetTextHandler = (condition, businessPage) =>
  condition ? businessPage.text[0].value : '';

export const enSetTitleHandler = (condition, businessPage) =>
  condition ? businessPage.title[1].value : '';

export const enSetTextHandler = (condition, businessPage) =>
  condition ? businessPage.text[1].value : '';

export const businessPageDispatchHandler = (
  mode,
  dispatch,
  update,
  add,
  updatePayload,
  addPayload
) => {
  if (mode) {
    return dispatch(update(updatePayload));
  }
  return dispatch(add(addPayload));
};

export const indexFinder = (i, filesArr, name, size) =>
  i === filesArr.findIndex((obj) => obj.name === name && obj.size === size);
