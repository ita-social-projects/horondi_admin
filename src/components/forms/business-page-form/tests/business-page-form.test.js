import React from 'react';
import * as reactRedux from 'react-redux';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import BusinessPageForm from '../index';
import { SaveButton } from '../../../buttons';
import LoadingBar from '../../../loading-bar';
import { mockBusinessPage, mockId } from './business-page-form.variables';

configure({ adapter: new Adapter() });

const mockSetFieldValue = jest.fn();
const mockSubmit = jest.fn();
const mockChange = jest.fn();
const mockBlur = jest.fn();

jest.mock('formik', () => ({
  ...jest.requireActual('formik'),
  useFormik: () => ({
    values: {},
    handleSubmit: mockSubmit,
    handleChange: mockChange,
    touched: {},
    errors: {},
    setFieldValue: mockSetFieldValue,
    handleBlur: mockBlur
  })
}));

jest.mock('react-router', () => ({
  withRouter: (component) => component
}));
jest.mock('react-router-dom', () => ({
  useHistory: () => ({ push: () => null })
}));
jest.mock('../../../../utils/business-page-form', () => ({
  setCodeHandler: jest.fn(() => null),
  uaSetTitleHandler: jest.fn(() => null),
  uaSetTextHandler: jest.fn(() => null),
  enSetTitleHandler: jest.fn(() => null),
  enSetTextHandler: jest.fn(() => null),
  businessPageDispatchHandler: jest.fn(() => null),
  indexFinder: jest.fn(() => null)
}));
jest.mock('../../../../utils/use-business-handlers', () => ({
  __esModule: true,
  default: () => ({
    code: 'some code',
    uaText: 'some text',
    uaTitle: 'some title',
    enText: 'some text',
    enTitle: 'some title',
    setCode: jest.fn(() => null),
    uaSetText: jest.fn(() => null),
    uaSetTitle: jest.fn(() => null),
    enSetText: jest.fn(() => null),
    enSetTitle: jest.fn(() => null),
    languages: ['ua', 'en'],
    createBusinessPage: jest.fn(() => null),
    files: ['image.png'],
    setFiles: jest.fn(() => null)
  })
}));
jest.mock('../../../../hooks/form-dialog/use-unsaved-changes-handler');
jest.mock('../../../../hooks/forms/use-changed-values-checker');
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useEffect: jest.fn((f) => f())
}));

describe('Straps form tests', () => {
  let spyOnUseSelector;
  let spyOnUseDispatch;
  let component;

  beforeEach(() => {
    spyOnUseSelector = jest
      .spyOn(reactRedux, 'useSelector')
      .mockImplementation(() => ({
        loading: false,
        businessPage: mockBusinessPage
      }));
    spyOnUseDispatch = jest.spyOn(reactRedux, 'useDispatch');
    spyOnUseDispatch.mockImplementation(() => jest.fn());

    component = shallow(<BusinessPageForm id={mockId} editMode />);
  });
  afterEach(() => {
    component.unmount();

    spyOnUseDispatch.mockClear();
    spyOnUseSelector.mockClear();
  });

  it('Should match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('Should simulate submit button', () => {
    component.find(SaveButton).prop('onClickHandler')();
    expect(mockSubmit).toHaveBeenCalled();
  });

  it('Should find loading bar', () => {
    spyOnUseSelector.mockImplementation(() => ({
      loading: true
    }));

    component = shallow(<BusinessPageForm id={mockId} editMode />);

    expect(component.find(LoadingBar)).toHaveLength(1);
  });

  it('Should click onsubmit button', () => {
    const event = {
      preventDefault: () => {}
    };

    jest.spyOn(event, 'preventDefault');

    component.find('form').simulate('submit', event);

    expect(event.preventDefault).toBeCalled();
  });
});
