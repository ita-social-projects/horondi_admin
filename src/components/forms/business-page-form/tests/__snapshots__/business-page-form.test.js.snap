// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Straps form tests Should match snapshot 1`] = `
<div
  className="makeStyles-container-5"
>
  <div
    className="makeStyles-buttonContainer-6"
  >
    <WithStyles(ForwardRef(Grid))
      className="makeStyles-fixedButtons-7"
      container={true}
      spacing={2}
    >
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-button-8"
        item={true}
      >
        <BackButton
          color="primary"
          initial={false}
          pathBack="/business-pages"
          type="button"
          variant="outlined"
        />
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        className="makeStyles-button-8"
        item={true}
      >
        <SaveButton
          color="primary"
          data-cy="save-btn"
          errors={Object {}}
          id="save"
          onClickHandler={[MockFunction]}
          title="Зберегти"
          type="submit"
          values={
            Object {
              "code": "some code",
              "enText": "some text",
              "enTitle": "some title",
              "uaText": "some text",
              "uaTitle": "some title",
            }
          }
        />
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </div>
  <div
    className="makeStyles-adminHeader-12"
  >
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-materialTitle-13"
      data-cy="add-header"
      variant="h1"
    >
      Додати нову бізнес сторінку
    </WithStyles(ForwardRef(Typography))>
  </div>
  <form
    onSubmit={[Function]}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={12}
    >
      <WithStyles(ForwardRef(Paper))
        className="makeStyles-businessPageForm-1"
      >
        <WithStyles(ForwardRef(TextField))
          className="makeStyles-textField-2"
          data-cy="page-code"
          id="code"
          label="Код сторінки"
          onBlur={[MockFunction]}
          onChange={[MockFunction]}
          value="some code"
          variant="outlined"
        />
      </WithStyles(ForwardRef(Paper))>
    </WithStyles(ForwardRef(Grid))>
    <LanguagePanel
      inputOptions={
        Object {
          "errors": Object {},
          "handleBlur": [MockFunction],
          "handleChange": [MockFunction],
          "inputs": Array [
            Object {
              "label": Object {
                "en": "Title",
                "ua": "Заголовок",
              },
              "name": "title",
              "required": true,
            },
            Object {
              "isEditor": true,
              "label": Object {
                "en": "Text",
                "ua": "Текст",
              },
              "name": "text",
              "setFiles": [MockFunction],
            },
          ],
          "setFieldValue": [MockFunction],
          "touched": Object {},
          "values": Object {
            "code": "some code",
            "enText": "some text",
            "enTitle": "some title",
            "uaText": "some text",
            "uaTitle": "some title",
          },
        }
      }
      key="ua"
      lang="ua"
    />
    <LanguagePanel
      inputOptions={
        Object {
          "errors": Object {},
          "handleBlur": [MockFunction],
          "handleChange": [MockFunction],
          "inputs": Array [
            Object {
              "label": Object {
                "en": "Title",
                "ua": "Заголовок",
              },
              "name": "title",
              "required": true,
            },
            Object {
              "isEditor": true,
              "label": Object {
                "en": "Text",
                "ua": "Текст",
              },
              "name": "text",
              "setFiles": [MockFunction],
            },
          ],
          "setFieldValue": [MockFunction],
          "touched": Object {},
          "values": Object {
            "code": "some code",
            "enText": "some text",
            "enTitle": "some title",
            "uaText": "some text",
            "uaTitle": "some title",
          },
        }
      }
      key="en"
      lang="en"
    />
  </form>
</div>
`;
